File: pages/refactorings/extract-method/02-running-balance/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Control.Monad (unless)
import System.Exit (exitFailure)
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import qualified Before
import qualified After

genAmount, genLimit, genFee :: Gen Int
genAmount = Gen.int (Range.linear (-30) 30)
genLimit  = Gen.int (Range.linear 0 20)
genFee    = Gen.int (Range.linear 1 20)

prop_settle_agrees :: Property
prop_settle_agrees = withTests 500 $ property $ do
  opening <- forAll genAmount
  limit   <- forAll genLimit
  fee     <- forAll genFee
  txs     <- forAll (Gen.list (Range.linear 0 20) genAmount)
  Before.settle opening limit fee txs === After.settle opening limit fee txs

prop_step_is_one_step :: Property
prop_step_is_one_step = withTests 500 $ property $ do
  balance <- forAll genAmount
  limit   <- forAll genLimit
  fee     <- forAll genFee
  tx      <- forAll genAmount
  After.step limit fee balance tx === Before.settle balance limit fee [tx]

main :: IO ()
main = do
  ok <- checkParallel $ Group "Props"
    [ ("settle: Before == After", prop_settle_agrees)
    , ("step is settle over a single transaction", prop_step_is_one_step)
    ]
  unless ok exitFailure

File: pages/refactorings/extract-method/02-running-balance/Spec.scala
//> using scala 3.3.4
//> using dep qa.hedgehog::hedgehog-core:0.14.0
//> using dep qa.hedgehog::hedgehog-runner:0.14.0
import hedgehog.*, hedgehog.core.*, hedgehog.runner.*

object Props extends Properties:
  def tests: List[Test] = List(
    property("settle: Before == After", settleAgrees).withTests(500),
  )

  val genAmount: Gen[Int] = Gen.int(Range.linear(-30, 30))
  val genLimit: Gen[Int]  = Gen.int(Range.linear(0, 20))
  val genFee: Gen[Int]    = Gen.int(Range.linear(1, 20))

  def settleAgrees: Property =
    for
      opening <- genAmount.forAll
      limit   <- genLimit.forAll
      fee     <- genFee.forAll
      txs     <- genAmount.list(Range.linear(0, 20)).forAll
    yield Before.settle(opening, limit, fee, txs) ====
      After.settle(opening, limit, fee, txs)

@main def spec(): Unit =
  val results = Props.tests.map { t =>
    val r = Property.check(t.withConfig(PropertyConfig.default),
      t.result, Seed.fromTime())
    println(Test.renderReport("Props", t, r,
      ansiCodesSupported = false))
    r.status
  }
  if !results.forall(_ == Status.ok) then sys.exit(1)
